prymer/api/picking.py

```diff
@@ -186,285 +79,109 @@ def score(
+    right_primers: list[Oligo],
     target: Span,
-    params: FilteringParams,
-    fasta: FastaFile,
-) -> list[PrimerPair]:
+    amplicon_sizes: MinOptMax[int],
+    amplicon_tms: MinOptMax[float],
+    max_heterodimer_tm: Optional[float],
+    weights: PrimerAndAmpliconWeights,
+    fasta_path: Path,
+) -> Iterator[PrimerPair]:
     """Builds primer pairs from individual left and primers.
 
+    Only primer pairs that meet the requirements for amplicon sizes and tms will be returned.
+    In addition, if `max_heterodimer_tm` is provided then primer pairs with a heterodimer tm
+    exceeding the maximum will also be discarded.
+
     Args:
         left_primers: the left primers
         right_primers: the right primers
         target: the genome mapping for the target
-        params: the parameters used for filtering
-        fasta: the FASTA file from which the amplicon sequence will be retrieved.
+        amplicon_sizes: minimum, optimal, and maximum amplicon sizes (lengths)
+        amplicon_tms: minimum, optimal, and maximum amplicon Tms
+        max_heterodimer_tm: if supplied, heterodimer Tms will be calculated for primer pairs,
+          and those exceeding the maximum Tm will be discarded
+        weights: the set of penalty weights
+        fasta_path: the path to the FASTA file from which the amplicon sequence will be retrieved.
 
     Returns:
-        the list of primer pairs, sorted by penalty (increasing)
+        an iterator over all the valid primer pairs, unsorted
     """
-    # generate all the primer pairs
-    primer_pairs: list[PrimerPair] = []
-    for left_primer in left_primers:
-        for right_primer in right_primers:
-            if left_primer.span.refname != right_primer.span.refname:
-                raise ValueError(
-                    "Cannot create a primer pair from left and right primers on different"
-                    f" references: left: '{left_primer.span.refname}'"
-                    f" right: '{right_primer.span.refname}'"
+    # Short circuit if we have no left primers or no right primers
+    if not any(left_primers) or not any(right_primers):
+        return
+
+    if not all(p.span.refname == target.refname for p in left_primers):
+        raise ValueError("Left primers exist on different reference than target.")
+
+    if not all(p.span.refname == target.refname for p in right_primers):
+        raise ValueError("Left primers exist on different reference than target.")
+
+    # Grab the sequence we'll use to fill in the amplicon sequence
+    with FastaFile(f"{fasta_path}") as fasta:
+        region_start = min(p.span.start for p in left_primers)
+        region_end = max(p.span.end for p in right_primers)
+        bases = fasta.fetch(target.refname, region_start - 1, region_end)
+
+    with NtThermoAlign() as ntthal:
+        # generate all the primer pairs that don't violate hard size and Tm constraints
+        for lp in left_primers:
+            for rp in right_primers:
```

Comment:
`itertools.product` would help you un-nest this a level:

https://docs.python.org/3/library/itertools.html#itertools.product